src/Avalonia.Themes.Fluent/Controls/SplitButton.xaml

```diff
@@ -0,0 +1,205 @@
+﻿<Styles xmlns="https://github.com/avaloniaui"
+        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+        x:CompileBindings="True">
+
+  <Design.PreviewWith>
+    <Border Padding="60">
+      <StackPanel>
+        <SplitButton Content="Hello World">
+          <SplitButton.Flyout>
+            <Flyout>Hello</Flyout>
+          </SplitButton.Flyout>
+        </SplitButton>
+        <ToggleSplitButton Content="Hello World">
+          <ToggleSplitButton.Flyout>
+            <Flyout>Hello</Flyout>
+          </ToggleSplitButton.Flyout>
+        </ToggleSplitButton>
+      </StackPanel>
+    </Border>
+  </Design.PreviewWith>
+
+  <Styles.Resources>
+    <x:Double x:Key="SplitButtonPrimaryButtonSize">32</x:Double>
+    <x:Double x:Key="SplitButtonSecondaryButtonSize">32</x:Double>
+    <StreamGeometry x:Key="SplitButtonSecondaryButtonIconGeometry">M1939 486L2029 576L1024 1581L19 576L109 486L1024 1401L1939 486Z</StreamGeometry>
+  </Styles.Resources>
+
+  <Style Selector="SplitButton">
+    <Setter Property="Background" Value="{DynamicResource SplitButtonBackground}" />
+    <Setter Property="Foreground" Value="{DynamicResource SplitButtonForeground}" />
+    <Setter Property="BorderBrush" Value="{DynamicResource SplitButtonBorderBrush}" />
+    <Setter Property="BorderThickness" Value="{DynamicResource SplitButtonBorderThemeThickness}" />
+    <Setter Property="HorizontalAlignment" Value="Left" />
+    <Setter Property="VerticalAlignment" Value="Center" />
+    <!--<Setter Property="UseSystemFocusVisuals" Value="True" />
+    <Setter Property="FocusVisualMargin" Value="-3" />-->
+    <Setter Property="KeyboardNavigation.IsTabStop" Value="True" />
+    <Setter Property="Focusable" Value="True" />
+    <Setter Property="Padding" Value="{DynamicResource ButtonPadding}" />
+    <Setter Property="CornerRadius" Value="{DynamicResource ControlCornerRadius}" />
+    <Setter Property="Template">
+      <Setter.Value>
+        <ControlTemplate>
+          <Grid>
```

Comment:
You removed the root border... that had setters applied to it that affected the overall look of the control. This applied below the two buttons.

Reply:
Yes. BorderBrush is now applied directly on the internal buttons, which makes it look better, when you apply click animation on them (border moves together with button, and not static). We still have click animation just like Fluent v1 had it, though maybe not the same.

Reply:
Ok, the only potential concern there is the inside borders of each button. Perhaps the Primary button should only have a border on the bottom, left and top while the secondary button a border on the bottom, right and top. This doesn't matter much for Fluent v1 which is in Avalonia now which doesn't really use borders. It can be addressed in the future whenever Fluent v2 style is adopted.

Reply:
I checked UWP Fluent v2. The border does seem to go around the entire control but not the middle separator. So your changes don't support this.

![image](https://user-images.githubusercontent.com/17993847/151249460-10ff1d2c-15a7-4e88-a0eb-f1c8abe64e35.png)

![image](https://user-images.githubusercontent.com/17993847/151249691-e5473b44-ed51-42ef-8226-a0b54ea34d78.png)

![image](https://user-images.githubusercontent.com/17993847/151249596-a1aaefed-aef6-4b53-b1c5-bb82abff65ea.png)

Reply:
When the enter button is pressed over the old control before your changes -- and in WinUI upstream -- the entire background updates: over both Primary and Secondary button. We need to maintain the ability to change the background for the entire control at once IMO. This is easier than setting both buttons separately and we probably need to add back a pressed state for the control itself.

Reply:
Current style is still compatible with "border around the entire control" by changing border thickness on one of the buttons.
But I see that additional layer of Border will simplify some custom styles without changing template.
It's not needed for current style, but if you want, feel free to add. Or I can do it later.

Reply:
I'll take a look at the Fluent v2 style. I'm guessing they use border differently and will attempt to copy that in with the changes you've made. Doing so will future-proof this area.

Reply:
@maxkatz6 

> Current style is still compatible with "border around the entire control" by changing border thickness on one of the buttons.

I don't see how. It can't account for the separator gap between buttons. As mentioned elsewhere:

>The PrimaryButton and SecondaryButton themselves are used only for hit targets basically. The actual style changes are applied to the PrimaryBackgroundRectangle, SecondaryBackgroundRectangle, Border (top) and RootGrid (bottom). Note that one Border is ON TOP in z-index and spans all 3 columns, the other RootGrid is on the bottom of the z-index and still spans all three columns. This means the Border/RootGrid itself goes around the entire control which is actually used in some states.
>
>Now, it is still possible to remove the PrimaryBackgroundRectangle and SecondaryBackgroundRectangle but the Buttons CANNOT handle the border. This still needs to be done for the entire control at once or it won't properly manage the separator gap. This means you have to add back a Border control. However, you can place it on top of the buttons by adding it at the end of the XAML (as you know). That should still allow you to use the buttons for changing primary/secondary backgrounds without having the extra rectangles.
>
>Note: Keep in mind that the entire control background also should change in some situations (instead of just Primary/Secondary button backgrounds). Obviously you can't use the same border for both border and background or it would entirely occlude the buttons due to z-index. This is why there is the RootBorder as well as the Border for actual borders on top. This will be discussed separately above.

So the top-most Border is needed. However, the RootBorder needs to come back here as well. There are situations where the entire background of the control should change color at once (all that logic was removed....). I think it needs to be added back:
 1. The RootBorder with overall control background spanning all 3 columns
 2. The code-behind logic to manage the pressed state for the entire control
 3. Selectors needs to be updated to manage the entire background in addition to the individual button backgrounds as well. You really tore into the selectors here and it's going to be tedious to bring it back.

All of this together means for external developers the SplitButton control can still somewhat be styled like a button without worrying about the separator gap as well.

I'll also leave this change to you if you don't mind.

Reply:
First of all, let's forget about v1 style. In WinUI 2 it doesn't even match Button style
![image](https://user-images.githubusercontent.com/3163374/152103453-06edde62-2ac5-4585-afcc-a1a054393fe5.png)

Reply:
> There are situations where the entire background of the control should change color at once

No need for any additional borders for that. You can once again change both button backgrounds with single selector, if SplitButton is pressed with a keyboard.

> Buttons CANNOT handle the border

Currently border and separator (as part of the buttons border) is managed by internal buttons without any issues. It also does have other advantages:
1. There is no cropping artifacts (as you have seen in another issue, anti-aliasing might be disabled for corner radius again, until it will be fixed on SkiaSharp side)
2. Colored border follows button press animation instead of ugly sticking on its place. It will be way harder to synchronize some kinds of animations for both border and single button. Note that button press animation exists only in FluentV1 and it doesn't work with SplitButton there. But exists in current PR implementation. I guess that's the reason why.
3. It's still easy to style both buttons separately, if developers need it. Though it's a weak reason.
And disadvantage:
1. Colored border will also color separator inside (comparing to WinUI). I did it intentionally (not to mention how it simplifies template), but I can imagine if some developers would not want it. Though I can also imagine the opposite.

Reply:
Now, if developer needs to modify separator between buttons (make it thinner or thicker) it can be done by modifying buttons border thickness, making it non-uniform, as you mentioned above. Empty spacing can be added with margin. It won't allow to change its color exclusively from outer border though. For that there should be additional border in between buttons. It can be added. **But.**

If developer needs to have common outer border (especially if they added empty spacing in between buttons or colored separated border with non-zero width), there should be yet another separated border (either common parent for both buttons or same-level child with higher zindex). It can be added. **But.**

**But** with current style implementation we need neither of these parts to be styled in special way. Developer can add any additional borders to the custom template and set internal buttons border thickness to 0 and add missing styles for these borders. We try to keep build in controls flexible, but without making them overcomplicated. We don't need it.

Reply:
Answering another possible issue, with common border it's not easy to style state when single button is pressed. I.e., border needs to know state of another control inside of the template. 
It **can** be solved with special pseudoclasses for SplitButton. But it's not needed for FluentV1 or FluentV2. 

There are bigger problems with ToggleSplitButton using styles from another control and non-reused Clicked event and ignoring code duplication with Button/ButtonBase. But we agreed to ignore these issues until customers will actually complain about it. In my opinion we should do the same here as well.

Upd: while writing it I understood that it actually can be solved with custom template, if developer adds it with border:
In developer's custom template it can be easier solved with code like:
```
<Border Classes.pressed-primary="{Binding #PART_PrimaryButton.IsPressed}"
              Classes.pressed-secondary="{Binding #PART_PrimaryButton.IsPressed}" />
```
Now these classes can be used in the styles.

Reply:
Ok, we need feedback from others here or other core maintainers. You and eye aren't going to see eye-to-eye.

I think what surprises me the most is how easily you throw away past precedent including decisions made by Microsoft's dev and design teams. They aren't perfect by any means but a lot more thought goes into things than you give them credit. Avalonia isn't so different to justify throwing everything away and starting over with a complete re-write.

The changes you made broke a few things.

 1. Try setting a border thickness of 2 to the control. This is reasonable in styling yet with your changes I now get a separator gap of 4. When a dev change the border of a control they don't want to change the separator distance between primary/secondary buttons... the control should style like a button.
 2. Using the borders as a separator also has the color concerns as you realize. The original design has a completely empty separator gap that shows through to the background color. You have removed this and now its the color of whatever the border brush is. That is an unwanted and frankly incorrect change.

>> There are situations where the entire background of the control should change color at once
>
> No need for any additional borders for that. You can once again change both button backgrounds with single selector, if SplitButton is pressed with a keyboard.

If I set the Background of the control to Red normally in XAML. Will I get red here including the separator? No. Fundamentally, I can no longer style the control like a button where background applies to the whole background including the separator gap and border applies to the entire outside border excluding the separator gap. I think this is a problem.

> Currently border and separator (as part of the buttons border) is managed by internal buttons without any issues.

Disagree there are no issues. Separator color and width are big ones.

> There is no cropping artifacts (as you have seen in another issue, anti-aliasing might be disabled for corner radius again, until it will be fixed on SkiaSharp side)

I already had removed cropping and was setting the corner radius of all controls. There should be no such issue doing this. If there is, that should be considered separately.

> Colored border follows button press animation instead of ugly sticking on its place. It will be way harder to synchronize some kinds of animations for both border and single button. Note that button press animation exists only in FluentV1 and it doesn't work with SplitButton there. But exists in current PR implementation. I guess that's the reason why.

This is the best argument I've heard for doing it the way you have. I don't think it outweighs the concerns though.

> It's still easy to style both buttons separately, if developers need it. Though it's a weak reason.

And it's harder to style the overall control as a button.

> But we agreed to ignore these issues until customers will actually complain about it. In my opinion we should do the same here as well.

The difference is I'm relying on past convention when I say we should leave things as-is. You are unilaterally going a new direction.

Reply:
> the control should style like a button

I think that's exact detail on which we can't agree. 
Let me explain. This control is a combination of two buttons. It's not just one button (developers always could always put a button inside of another button to achieve similar to that, I did it before). What exactly is needed - decides developer (and UX/UI team). Not we two. Not Microsoft.
As you told above developers might want to style it as a single button. But other developers might want to style it as two buttons:
![image](https://user-images.githubusercontent.com/3163374/152278816-c52dc4ab-0b52-46f4-8482-6e2ae3b49902.png)
And here is the problem. With common border hacks from UWP we can achieve single button styling easily, correct. But there is no way to achieve separated buttons styling. Of course, with custom template it's possible, but we already wrote couple of pages of text above only about default template being enough flexible.

> what surprises me the most is how easily you throw away past precedent including decisions made by Microsoft's dev 

You already know that styling system in Avalonia has different capabilities. You can do things that weren't easily possible in UWP (flexible selectors styling), as well as opposite is true also (you can't have state management as it is in UWP, without writing hundreds of non-optimal styles).
And I believe you also noticed that FluentV1 and FluentV2 winui styles for SplitButton aren't perfect either. And for any customization you need to replace template (using your examples, try to set border thickness to 8 and set border brush on FluentV2, you will see broken control because of common border being on top of both buttons).
I understand you want to see Avalonia closer in behavior and controls-set to what UWP has. But that never was a goal for Avalonia. It always nice to have similarities in frameworks though, so simplify it for new users.

> Will I get red here including the separator? No.

As it won't in FluentV1 - separator is transparent (and no, you can't see button's background behind it). As you can do it in both WPF and Avalonia with custom template.

Reply:
> Try setting a border thickness of 2 to the control

That's a good argument. We can go to a compromise and add border in between and adjust border thickness of elements with converters (we already have one build in for that). It will work well with current style, won't break click animation, still allow to style buttons separately and also allow more freedom with styling separated independently.
It is a bit ugly, but still better than UWP has with common border on top of everything.
This approach has a visual difference from FluentV1/FluentV2, where button pointerover/pressed affects common border instead of buttons separately. 

https://user-images.githubusercontent.com/3163374/152281570-c60a5e7d-747e-4810-9c4f-67a43c4931bc.mp4

Reply:
> I think that's exact detail on which we can't agree.

Yes, fundamentally that's it

> As you told above developers might want to style it as a single button. But other developers might want to style it as two buttons:

Yes, but WinUI already has established convention. Why change the default assumption from what works today? Your argument that it prevents styling as two buttons and also that it doesn't work well with press transformations is partly valid of course. However, no matter what there are going to be trade-offs. This was all discussed in the past before our pages of discussion here. That is re-inventing the wheel so to speak which I think is an unnecessary use of time. But discussions of opinions are important...

> You already know that styling system in Avalonia has different capabilities. 

Nothing we discussed here in this thread is fundamentally different between WinUI and Avalonia. There is nothing preventing the template in Avalonia from being almost the same in implementation and exactly the same in function. 

> (using your examples, try to set border thickness to 8 and set border brush on FluentV2, you will see broken control because of common border being on top of both buttons).

I'm not sure what exactly would be broken in this case, but a Border thickness of 8 is extreme. It's not worth my time to test right now.

> As it won't in FluentV1 - separator is transparent (and no, you can't see button's background behind it). As you can do it in both WPF and Avalonia with custom template.

In WinUI the SplitButton background is set to the root grid. Then the separator is an empty column of the grid between the two buttons. How would you not be able to see the Background in this case? Of course I am just reading the template and won't spend the time to test; however, it seems clear.

> We can go to a compromise and add border in between and adjust border thickness of elements with converters (we already have one build in for that).

I think there was still an option to re-architect the press transform animation to apply to the entire control instead of each internal button individually. However, this idea for a compromise is functional so I'll agree to it. You know my thinking otherwise. I have the same feeling that this design is a bit ugly as you said. It also has some downsides.

The button press animation creates an additional gap between the primary/secondary button border and the separator border. Visually, this gap does not look good and it breaks the border around the control when you press (you can test this out in the ControlCatalog with the new Re-themed SplitButton example). This is where a uniform outside border would look good. This issue should not exist in standard Fluent themes though:
 1) In Fluent v1 there is no border by default so this issue is not visible
 2) In any future Fluent v2 the Button press transform animation doesn't exist so no border gap would occur

Also, the use of converters so many places in a default style is generally frowned upon on my end.

Anyway, you know well my standpoint on this and the compromise is functional. We'll wait for any additional feedback in real-world usage. Marking resolved.